cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -0,0 +1,432 @@
+   * ascending/descending order of each column in a row. If empty, it is assumed all columns are
+   * sorted in ascending order.
+   * @param null_precedence Optional, device array the same length as a row and indicates how null
+   * values compare to all other for every column. If it is nullptr, then null precedence would be
+   * `null_order::BEFORE` for all columns.
+   * @param stream The stream to launch kernels and h->d copies on while preprocessing.
+   */
+  preprocessed_table(table_view const& table,
+                     host_span<order const> column_order,
+                     host_span<null_order const> null_precedence,
+                     rmm::cuda_stream_view stream);
+
+  // TODO: Should we add a static create method that returns a shared_ptr?
+
+ private:
+  friend class self_comparator;
+
+  /**
+   * @brief Implicit conversion operator to a `table_device_view` of the preprocessed table.
+   *
+   * @return table_device_view
+   */
+  operator table_device_view() { return **d_t; }
+
+  /**
+   * @brief Get a device array containing the desired order of each column in the preprocessed table
+   *
+   * @return std::optional<device_span<order const>> Device array containing respective column
+   * orders. If no explicit column orders were specified during the creation of this object then
+   * this will be `nullopt`.
+   */
+  [[nodiscard]] std::optional<device_span<order const>> column_order() const
+  {
+    return d_column_order.size() ? std::optional<device_span<order const>>(d_column_order)
+                                 : std::nullopt;
+  }
+
+  /**
+   * @brief Get a device array containing the desired null precedence of each column in the
+   * preprocessed table
+   *
+   * @return std::optional<device_span<null_order const>> Device array containing respective column
+   * null precedence. If no explicit column null precedences were specified during the creation of
+   * this object then this will be `nullopt`.
+   */
+  [[nodiscard]] std::optional<device_span<null_order const>> null_precedence() const
+  {
+    return d_null_precedence.size()
+             ? std::optional<device_span<null_order const>>(d_null_precedence)
+             : std::nullopt;
+  }
+
+  /**
+   * @brief Get a device array containing the depth of each column in the preprocessed table
+   *
+   * A `depth` value is the number of nested levels as parent of the column in the original,
+   * non-preprocessed table, which were pruned during preprocessing.
+   * For example, if the original table has a column `Struct<Struct<int, float>, decimal>`,
+   *      S
+   *     / \
+   *    S   d
+   *   / \
+   *  i   f
+   * then after preprocessing, we get three columns:
+   * `Struct<Struct<int>>`, `float`, and `decimal`.
+   * 0   2   1  <- depths
+   * S
+   * |
+   * S       d
+   * |
+   * i   f
+   * The depth of the first column is 0 because it contains all its parent levels, while the depth
```

Comment:
The first column isn't`i`. It's `S<S<i>>`. And that top level `S` has a depth 0. 

> Is this structure data required to assemble it again on the other end?

Sort of. This structure will be used to determine which columns to skip in case of a null e.g if there is a null at the second `S` in `S<S<i>>`, then the null appears at depth 1. Since `f` is at depth 2, it must be a child of that `S` in the original column. So skip it. `d` has a depth 1 which means `d` is a sibling of the second `S` and so the latter's nullness has no effect on it.